akka-stream/src/main/scala/akka/stream/KillSwitch.scala

```diff
@@ -145,6 +145,38 @@ trait KillSwitch {
 }
 //#kill-switch
 
+private[stream] final class TerminationSignal {
+  final class Listener private[TerminationSignal] {
+    private[TerminationSignal] val promise = Promise[Done]
+    def future: Future[Done] = promise.future
+    def unregister(): Unit = removeListener(this)
+  }
+
+  private[this] val lock = new Object
+  private[this] var _listeners = Set.empty[Listener]
+  private[this] var _completedWith: Option[Try[Done]] = None
+
+  def broadcast(result: Try[Done]): Unit = lock.synchronized {
+    if (_completedWith.isEmpty) {
+      _completedWith = Some(result)
+      for (listener ← _listeners) listener.promise.tryComplete(result)
+    }
+  }
+
+  def createListener(): Listener = lock.synchronized {
+    val listener = new Listener
+    _completedWith match {
+      case Some(result) ⇒ listener.promise.complete(result)
+      case None         ⇒ _listeners += listener
+    }
+    listener
+  }
+
+  private def removeListener(listener: Listener): Unit = lock.synchronized {
+    _listeners -= listener
+  }
+}
```

Comment:
let's see if we can write this without `synchronized`, perhaps something like this:

```scala
private[stream] final class TerminationSignal {
  final class Listener private[TerminationSignal] {
    private[TerminationSignal] val promise = Promise[Done]
    def future: Future[Done] = promise.future
    def unregister(): Unit = removeListener(this)
  }

  private[this] var _listeners = new AtomicReference[Set[Listener]](Set.empty[Listener])
  private[this] val _completedWith = new AtomicReference[Option[Try[Done]]](None)

  def broadcast(result: Try[Done]): Unit = {
    if (_completedWith.compareAndSet(None, Some(result))) {
      for (listener ← _listeners.get) listener.promise.tryComplete(result)
    }
  }

  @tailrec final def createListener(): Listener = {
    val oldListeners = _listeners.get
    val listener = new Listener
    if (_listeners.compareAndSet(oldListeners, oldListeners + listener)) {
      _completedWith.get match {
        case Some(result) => listener.promise.tryComplete(result) // ok to tryComplete several times
        case None         => // not completed yet
      }
      listener
    } else
      createListener() // recursive, CAS failed
  }

  @tailrec private def removeListener(listener: Listener): Unit = {
    val oldListeners = _listeners.get
    if (!_listeners.compareAndSet(oldListeners, oldListeners - listener))
      removeListener(listener) // recursive, CAS failed
  }
}
```

Reply:
I was thinking to use a concurrent non-blocking collection like `scala.collection.concurrent.TrieMap` in order to reduce complexity.

There is a bit of a race condition around reading the completion state and modifying the collection. In the suggested implementation we would be always inserting into the collection, but the collection is emptied only once. So we have a memory leak again. We should use the `check condition -> update -> re-check condition` pattern to remedy that.

Reply:
After some more thinking about this, the `postStop()` hook should trigger the appropriate collection cleanup, so there should be no memory leak even with unconditional insert into `_listeners`.